python/dgl/distributed/client.py

```diff
@@ -2,33 +2,155 @@
+        ip = data[1]
+        if ip in local_ip4_addr_list():
+            res = machine_id
+            break
+    return res
+
+def get_local_usable_addr():
+    """Get local usable IP and port
+
+    Returns
+    -------
+    str
+        IP address, e.g., '192.168.8.12:50051'
+    """
+    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    try:
+        # doesn't even have to be reachable
+        s.connect(('10.255.255.255', 1))
+        ip_addr = s.getsockname()[0]
+    except:
+        ip_addr = '127.0.0.1'
+    finally:
+        s.close()
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    s.bind(("", 0))
+    s.listen(1)
+    port = s.getsockname()[1]
+    s.close()
+
+    return ip_addr + ':' + str(port)
+
+def connect_to_server(ip_config, queue_size=20*1024*1024*1024, net_type='socket'):
     """Connect this client to server.
 
     Parameters
     ----------
-    server_namebook : ...
-        Server names.
+    ip_config : str
+        Path of server IP configuration file.
+    queue_size : int
+        Size (bytes) of client queue buffer (~20 GB on default).
+        Note that the 20 GB is just an upper-bound and DGL uses zero-copy and
+        it will not allocate 20GB memory at once.
+    net_type : str
+        networking type, e.g., 'socket' (on default) or 'mpi' (do not support yet).
 
     Raises
     ------
     ConnectionError : If anything wrong with the connection.
     """
-    pass
+    assert queue_size > 0, 'queue_size (%d) cannot be a negative number.' % queue_size
+    assert net_type in ('socket', 'mpi'), \
+    'net_type (%s) can only be \'socket\' or \'mpi\'.' % net_type
+    # Register some basic service
+    rpc.register_service(rpc.CLIENT_REGISTER,
+                         rpc.ClientRegisterReuqest,
+                         rpc.ClientRegisterResponse)
+    rpc.register_service(rpc.SHUT_DOWN_SERVER,
+                         rpc.ShutDownRequest,
+                         None)
+    server_namebook = rpc.read_ip_config(ip_config)
+    num_servers = len(server_namebook)
+    rpc.set_num_server(num_servers)
+    group_count = []
+    max_machine_id = 0
+    for server_info in server_namebook.values():
+        group_count.append(server_info[3])
+        if server_info[0] > max_machine_id:
+            max_machine_id = server_info[0]
+    num_machines = max_machine_id+1
+    machine_id = get_local_machine_id(server_namebook)
+    rpc.create_sender(queue_size, net_type)
+    rpc.create_receiver(queue_size, net_type)
+    # Get connected with all server nodes
+    for server_id, addr in server_namebook.items():
+        server_ip = addr[1]
+        server_port = addr[2]
+        rpc.add_receiver_addr(server_ip, server_port, server_id)
+    rpc.sender_connect()
+    # Get local usable IP address and port
+    ip_addr = get_local_usable_addr()
+    client_ip, client_port = ip_addr.split(':')
+    # Register client on server
+    # a temp ID because we don't assign client ID yet
+    rpc.set_rank(0)
```

Comment:
why is rank always 0?

Reply:
Because in the beginning, the client doesn't have the ID and it is assigned by the server. 0 is temp ID to connect with server for some initialization.